editor/src/components/canvas/canvas-strategies/interaction-state.ts

```diff
@@ -0,0 +1,321 @@
+        modifiers: modifiers,
+      },
+      activeControl: currentState.activeControl,
+      sourceOfUpdate: currentState.activeControl,
+      lastInteractionTime: Date.now(),
+      userPreferredStrategy: currentState.userPreferredStrategy,
+      startedAt: currentState.startedAt,
+      globalTime: Date.now(),
+    }
+  } else {
+    return {
+      ...currentState,
+      globalTime: Date.now(),
+    }
+  }
+}
+
+export function strategySwitchInteractionDataReset(interactionData: InputData): InputData {
+  switch (interactionData.type) {
+    case 'DRAG':
+      if (interactionData.drag == null || interactionData.prevDrag == null) {
+        return interactionData
+      } else {
+        return {
+          ...interactionData,
+          dragStart: offsetPoint(interactionData.dragStart, interactionData.prevDrag),
+          drag: pointDifference(interactionData.prevDrag, interactionData.drag),
+          prevDrag: null,
+        }
+      }
+    case 'KEYBOARD':
+      return interactionData
+    default:
+      const _exhaustiveCheck: never = interactionData
+      throw new Error(`Unhandled interaction type ${JSON.stringify(interactionData)}`)
+  }
+}
+
+// Hard reset means we need to ignore everything happening in the interaction until now, and replay all the dragging
+export function interactionDataHardReset(interactionData: InputData): InputData {
+  switch (interactionData.type) {
+    case 'DRAG':
+      if (interactionData.drag == null) {
+        return interactionData
+      } else {
+        const currentDrag = interactionData.drag ?? zeroCanvasPoint
+        return {
+          ...interactionData,
+          dragStart: interactionData.originalDragStart,
+          drag: pointDifference(
+            interactionData.originalDragStart,
+            offsetPoint(interactionData.dragStart, currentDrag),
+          ),
+        }
+      }
+    case 'KEYBOARD':
+      return interactionData
+    default:
+      const _exhaustiveCheck: never = interactionData
+      throw new Error(`Unhandled interaction type ${JSON.stringify(interactionData)}`)
+  }
+}
+
+export function strategySwitchInteractionSessionReset(
+  interactionSession: InteractionSession,
+): InteractionSession {
+  return {
+    ...interactionSession,
+    interactionData: strategySwitchInteractionDataReset(interactionSession.interactionData),
+  }
+}
+
+// Hard reset means we need to ignore everything happening in the interaction until now, and replay all the dragging
+export function interactionSessionHardReset(
+  interactionSession: InteractionSession,
+): InteractionSession {
+  return {
+    ...interactionSession,
+    interactionData: interactionDataHardReset(interactionSession.interactionData),
+  }
+}
+
+export function hasModifiersChanged(
+  prevInteractionData: InputData | null,
+  interactionData: InputData | null,
+): boolean {
+  return (
+    interactionData?.type === 'DRAG' &&
+    prevInteractionData?.type === 'DRAG' &&
```

Comment:
If this is only for dragging, then the function name should reflect that - or maybe only allow DragInteractionData parameters?

Reply:
renamed